### PR TITLE
Add Gitee provider

### PR DESCRIPTION
**Pull request by HahaMango**
Hi, I implemented a gitee provider.

The provider oAuth workflow is basically the same as github, and i also try to keep the code consistent with other providers.

I also wrote test code for the provider and passed the test.

How to use:

```csharp
builder.AddGitee(options => 
{
    options.ClientId = "your client Id ";
    options.ClientSecret = "your client secret ";
});
```

### CODE REVIEW

**Comment by martincostello** [comment]
Thanks for the Pull Request @HahaMango - I'll try and look at this sometime in the next few days.

**Comment by HahaMango** [comment]
OK! @martincostello

File: src/AspNet.Security.OAuth.Gitee/GiteeAuthenticationHandler.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                                /* Status: */ response.StatusCode,
+                                /* Headers: */ response.Headers.ToString(),
+                                /* Body: */ await response.Content.ReadAsStringAsync());
+
+                throw new HttpRequestException("An error occurred while retrieving the user profile.");
+            }
+
+            var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Gitee/GiteeAuthenticationHandler.cs]:
```suggestion
            using var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
```

File: src/AspNet.Security.OAuth.Gitee/GiteeAuthenticationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            [NotNull] ILoggerFactory logger,
+            [NotNull] UrlEncoder encoder,
+            [NotNull] ISystemClock clock)
+            : base(options, logger, encoder, clock)
+            {
+            }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync(ClaimsIdentity identity, AuthenticationProperties properties, OAuthTokenResponse tokens)
+        {
+            var request = new HttpRequestMessage(HttpMethod.Get, Options.UserInformationEndpoint);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Gitee/GiteeAuthenticationHandler.cs]:
```suggestion
            using var request = new HttpRequestMessage(HttpMethod.Get, Options.UserInformationEndpoint);
```

File: src/AspNet.Security.OAuth.Gitee/GiteeAuthenticationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+
+            var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Gitee/GiteeAuthenticationHandler.cs]:
```suggestion
            using var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
```

File: src/AspNet.Security.OAuth.Gitee/GiteeAuthenticationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                }
+            }
+
+            await Options.Events.CreatingTicket(context);
+            return new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
+        }
+
+        protected async Task<string> GetEmailAsync([NotNull] OAuthTokenResponse tokens)
+        {
+            var request = new HttpRequestMessage(HttpMethod.Get, Options.UserEmailsEndpoint);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Gitee/GiteeAuthenticationHandler.cs]:
```suggestion
            using var request = new HttpRequestMessage(HttpMethod.Get, Options.UserEmailsEndpoint);
```

File: src/AspNet.Security.OAuth.Gitee/GiteeAuthenticationHandler.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+
+            var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Gitee/GiteeAuthenticationHandler.cs]:
```suggestion
            using var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
```

File: src/AspNet.Security.OAuth.Gitee/GiteeAuthenticationHandler.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                Logger.LogWarning("An error occurred while retrieving the email address associated with the logged in user: " +
+                                  "the remote server returned a {Status} response with the following payload: {Headers} {Body}.",
+                                  /* Status: */ response.StatusCode,
+                                  /* Headers: */ response.Headers.ToString(),
+                                  /* Body: */ await response.Content.ReadAsStringAsync());
+
+                throw new HttpRequestException("An error occurred while retrieving the email address associated to the user profile.");
+            }
+
+            var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Gitee/GiteeAuthenticationHandler.cs]:
```suggestion
            using var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
```

File: src/AspNet.Security.OAuth.Gitee/GiteeAuthenticationOptions.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    /// <summary>
+     /// Defines a set of options used by <see cref="GiteeAuthenticationHandler"/>.
+     /// </summary>
+    public class GiteeAuthenticationOptions : OAuthOptions
+    {
+        public GiteeAuthenticationOptions()
+        {
+            ClaimsIssuer = GiteeAuthenticationDefaults.Issuer;
+
+            CallbackPath = new PathString(GiteeAuthenticationDefaults.CallbackPath);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Gitee/GiteeAuthenticationOptions.cs]:
```suggestion
            CallbackPath = GiteeAuthenticationDefaults.CallbackPath;
```

**Review by martincostello — COMMENTED** [review]
Just a few minor alterations needed.

**Review by martincostello — COMMENTED** [review]
Also the provider needs adding to the README.

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
Thanks for your contribution @HahaMango 🎉

**Comment by martincostello** [comment]
This provider is now available from NuGet.org: https://www.nuget.org/packages/AspNet.Security.OAuth.Gitee/3.1.0

**Pull request merged**